okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -28,6 +29,135 @@ private DepResolution() {
+        } else {
+          list.addAll(res);
+        }
+      }
+    }
+    if (list.isEmpty()) {
+      return "";
+    } else {
+      return String.join("\n", list);
+    }
+  }
+
+  /**
+   * Test interfaces for a set of enabled modules.
+   * @param modsEnabled enabled modules
+   * @return list of errors ; empty list if not error(s)
+   */
+  public static List<String> checkEnabled(Map<String, ModuleDescriptor> modsEnabled) {
+    return interfaceCheck(modsEnabled, modsEnabled, Collections.emptyList(), false,
+        new HashSet<>());
+  }
+
+  /**
+   * Install modules with dependency checking only.
+   * @param modsAvailable available modules
+   * @param modsEnabled enabled modules (for some tenant)
+   * @param tml install list with actions
+   * @param reinstall whether to re-install
+   */
+  public static void install(Map<String, ModuleDescriptor> modsAvailable,
+      Map<String, ModuleDescriptor> modsEnabled, List<TenantModuleDescriptor> tml,
+      boolean reinstall) {
+    installMaxIterations(modsAvailable, modsEnabled, tml, reinstall, 100);
+  }
+
+  /**
+   * Return top-N set of modules - in order of module ID.
+   * @param limit max number for each module (Top-N)
+   * @param mdl modules to consider (will be modified!)
+   */
+  public static void getLatestProducts(int limit, List<ModuleDescriptor> mdl) {
+    mdl.sort(Collections.reverseOrder());
+    Iterator<ModuleDescriptor> it = mdl.listIterator();
+    String product = "";
+    int no = 0;
+    while (it.hasNext()) {
+      ModuleDescriptor md = it.next();
+      if (!product.equals(md.getProduct())) {
+        product = md.getProduct();
+        no = 0;
+      } else if (no >= limit) {
+        it.remove();
+      }
+      no++;
+    }
+  }
+
+  static boolean moduleDepProvided(List<ModuleDescriptor> modules, ModuleDescriptor md) {
+    List<InterfaceDescriptor> requiresOptionalList = md.getRequiresOptionalList();
+    for (InterfaceDescriptor req : requiresOptionalList) {
+      boolean found = false;
+      for (ModuleDescriptor md1 : modules) {
+        InterfaceDescriptor[] providesList = md1.getProvidesList();
+        for (InterfaceDescriptor prov : providesList) {
+          if (prov.isCompatible(req)) {
+            found = true;
+          }
+        }
+      }
+      if (!found) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  static void topoSort(List<ModuleDescriptor> modules) {
+    List<ModuleDescriptor> result = new LinkedList<>();
+
+    boolean more = true;
+    while (more) {
+      more = false;
+      Iterator<ModuleDescriptor> iterator = modules.iterator();
+      while (iterator.hasNext()) {
+        ModuleDescriptor md = iterator.next();
+        if (moduleDepProvided(result, md)) {
+          result.add(md);
+          iterator.remove();
+          more = true;
+        }
+      }
+    }
+    // add remaining result to modules (such as modules with optional interfaces requirements
```

Comment:
Multiple modules with unsatisfied optional dependency can be sorted correctly if all optional dependencies that cannot satisfied are removed.

Reply:
Now done, but not by removal, but by checking all interfaces provided (if not part of that, the optional interface is ok)